telluric/georaster.py

```diff
@@ -108,14 +111,87 @@ def merge_all(rasters, roi=None, dest_resolution=None, merge_strategy=MergeStrat
         roi, resolution=dest_resolution, band_names=rasters[0].band_names,
         dtype=rasters[0].dtype, shape=shape, ul_corner=ul_corner, crs=crs)
 
-    projected_rasters = [_prepare_other_raster(empty, raster) for raster in rasters]
-    projected_rasters = [raster for raster in projected_rasters if raster is not None]
-    func = partial(_merge, merge_strategy=merge_strategy)  # type: Callable[[_Raster, _Raster], _Raster]
-    raster = reduce(func, projected_rasters, empty)
+    # Create a list of single band rasters
+    projected_rasters = _prepare_rasters(rasters, merge_strategy, empty)
+
+    # Merge common bands
+    projected_rasters = _merge_common_bands(projected_rasters)
```

Comment:
This makes the assumption that the input rasters have one mask per dataset. We could do better and make the code work in more general cases, but we should move the mask arithmetic to the beginning.